airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -66,6 +66,17 @@
  * <p>
  * All other message types are ignored.
  * </p>
+ *
+ * <p>
+ * Throughout the lifecycle of the consumer, records get promoted from buffered to flushed to
+ * committed. A record when it is received is immediately buffered. When the buffer fills up, all
+ * buffered records are flushed out of memory using the user-provided recordWriter. When this flush
+ * happens, a state message is moved from pending to flushed. On close, if the user-provided onClose
+ * function is successful, then the flushed state record is considered committed and is then
+ * emitted. We expect this class to only ever emit either 1 state message (in the case of a full or
```

Comment:
shouldn't this class be emitting state messages potentially continuously to allow halfway failure checkpointing?

Reply:
or is the hangup here on normalization?

Reply:
Normalization is orthogonal to this.

Emitting state as we go would be another valid strategy for approaching checkpointing in destinations. 

The current strategy is to consume records and put them in a temporary table in the destination until something does wrong. As soon as something does wrong (or we consume all records) attempt to commit everything in the temporary table into the final tables. This current approach gets us checkpointing behavior, but it is vulnerable to the case where the commit would have been successful for a subset of the flushed records but not all of them. I'm not sure to what extent this is really a high frequency failure point.

The reason to not go for emitting state as we go is just expedience. If we emit state as we go then we will need to be committing records from the temporary tables to the final tables as we go. So now each sync can potentially have multiple temporary tables. It is definitely doable and something we should probably shoot for, but at the time I didn't have appetite for the added complexity.